### PR TITLE
Auto Reset for .NET

### PR DESCRIPTION
**Pull request by MongoCaleb**
### Jira

https://jira.mongodb.org/browse/DOCSP-19827

### Staged Changes (Requires MongoDB Corp SSO)
https://docs-mongodbcom-staging.corp.mongodb.com/realm/docsworker-xlarge/19827/sdk/dotnet/advanced-guides/client-reset/

### CODE REVIEW

File: source/sdk/dotnet/advanced-guides/client-reset.txt [hunk 1]
```diff
@@ -14,32 +14,81 @@ Client Resets - .NET SDK
    :class: singlecol
 
 When using :doc:`{+sync+} </sync>`, it is possible that your client app will 
-need to handle a **client reset**. While not common, it can be caused when Sync was 
-terminated on the server. For example, a developer deploys a change that includes 
-a destructive schema change. To fix this, they must stop Sync, modify the schema, 
-and then re-enable Sync. This process causes the server to send a client reset 
-message to the clients.
+need to handle a **client reset**. While not a common event, it can be caused 
+when Sync was terminated on the server. For example, a developer deploys a
```
Comment (nathan-contino-mongo) [inline source/sdk/dotnet/advanced-guides/client-reset.txt:18]: Could we rephrase this to active voice?

File: source/sdk/dotnet/advanced-guides/client-reset.txt [hunk 2]
```diff
@@ -14,32 +14,81 @@ Client Resets - .NET SDK
...
+when Sync was terminated on the server. For example, a developer deploys a 
+change that includes a destructive schema change. To fix this, you must stop 
+Sync, modify the schema, and then re-enable Sync. This process causes the 
+server to send a client reset message to the clients.
 
 By default, the SDK handles client reset messages automatically, but only on app 
-startup. Until then, users will not receive any data from the server, and any 
-changes they make will be lost during the reset. Therefore, you should handle 
-client resets in your code to ensure users get up-to-date data. 
+startup. Until the app is restarted, users will not receive any new data from the
```
Comment (nathan-contino-mongo) [inline source/sdk/dotnet/advanced-guides/client-reset.txt:24]: More passive voice, rephrase to active maybe?

File: source/sdk/dotnet/advanced-guides/client-reset.txt [hunk 3]
```diff
@@ -14,32 +14,81 @@ Client Resets - .NET SDK
...
+change that includes a destructive schema change. To fix this, you must stop 
+Sync, modify the schema, and then re-enable Sync. This process causes the 
+server to send a client reset message to the clients.
 
 By default, the SDK handles client reset messages automatically, but only on app 
-startup. Until then, users will not receive any data from the server, and any 
-changes they make will be lost during the reset. Therefore, you should handle 
-client resets in your code to ensure users get up-to-date data. 
+startup. Until the app is restarted, users will not receive any new data from the 
+server, and any changes they make will be lost when the app is restarted and the
```
Comment (nathan-contino-mongo) [inline source/sdk/dotnet/advanced-guides/client-reset.txt:25]: "app is restarted" is an uncommon phrase and passive to boot, I think we can spice this up a little.

File: source/sdk/dotnet/advanced-guides/client-reset.txt [hunk 4]
```diff
@@ -14,32 +14,81 @@ Client Resets - .NET SDK
...
    changes made since the client last successfully synced.
 
-Example
--------
-When a reset is required, {+realm+} sends a  
-:dotnet-sdk:`ClientResetException <reference/Realms.Sync.Exceptions.ClientResetException.html>`. 
-You can add a listener on the 
-:dotnet-sdk:`Session.Error <reference/Realms.Sync.Session.html#Realms_Sync_Session_Error>` 
-event and handle the client reset by calling the 
+.. versionadded:: 10.11.1
```
Comment (nathan-contino-mongo) [inline source/sdk/dotnet/advanced-guides/client-reset.txt:34]: Personally I feel that this "version added" note on the rendered page is confusing. It's both rendered redundant by the very next sentence... and unclear in scope. Maybe we could just remove it, since versionadded doesn't let us explain what's actually new and what's old (and you already do that in the next sentence)?

File: source/sdk/dotnet/advanced-guides/client-reset.txt [hunk 5]
```diff
@@ -14,32 +14,81 @@ Client Resets - .NET SDK
...
+reset strategy in your :dotnet-sdk:`FlexibleSyncConfiguration 
+<reference/Realms.Sync.FlexibleSyncConfiguration.html>`. The 
+:dotnet-sdk:`ClientResetHandler<TBD>` property can be set to either the 
+:dotnet-sdk:`DiscardLocalResetHandler<TBD>` method or the 
+:dotnet-sdk:`ManualRecoveryHandler<TBD>` method. If you do not specify a 
+``ClientResetHandler``, the default is ``ManualRecoveryHandler``.  
+
+Using the DiscardLocalResetHandler
+----------------------------------
+When you specify a ``DiscardLocalResetHandler``, Realm
```
Comment (nathan-contino-mongo) [inline source/sdk/dotnet/advanced-guides/client-reset.txt:46]:
```suggestion
When you specify a ``DiscardLocalResetHandler``, the SDK 
```
Since the "Realm" term is ambiguous, would it make more sense to just say "the SDK" here?

File: source/sdk/dotnet/advanced-guides/client-reset.txt [hunk 6]
```diff
@@ -14,32 +14,81 @@ Client Resets - .NET SDK
...
+
+Be aware that this strategy *permanently deletes* all local unsynced changes 
+made since the last successful sync. To ensure local changes are incorporated,
+specify the :dotnet-sdk:`OnBeforeReset<TBD>` and
+:dotnet-sdk:`OnAfterReset<TBD>` callback methods. Within the ``OnBeforeReset`` 
+callback method, you can copy the existing data to a temporary realm, and then 
+manually merge the data into the reset realm in the ``OnAfterReset`` callback 
+method.
+
+If the client reset operation cannot complete when using the
```
Comment (nathan-contino-mongo) [inline source/sdk/dotnet/advanced-guides/client-reset.txt:60]: Would it be helpful to specify *why* the operation might not be able to complete here? (as far as I know, only a destructive schema change?)

**Review by nathan-contino-mongo — APPROVED** [review]
LGTM with some nitpicks

File: source/sdk/dotnet/advanced-guides/client-reset.txt [hunk 7]
```diff
@@ -14,32 +14,81 @@ Client Resets - .NET SDK
...
-Example
--------
-When a reset is required, {+realm+} sends a  
-:dotnet-sdk:`ClientResetException <reference/Realms.Sync.Exceptions.ClientResetException.html>`. 
-You can add a listener on the 
-:dotnet-sdk:`Session.Error <reference/Realms.Sync.Session.html#Realms_Sync_Session_Error>` 
-event and handle the client reset by calling the 
+.. versionadded:: 10.11.1
+
+Starting in version 10.11.1, the .NET SDK provides the option to specify a client
```
Comment (nirinchev) [inline source/sdk/dotnet/advanced-guides/client-reset.txt:36]: FYI this version is not set in stone as we're still waiting on a Cloud fix before we do a release.

File: source/sdk/dotnet/advanced-guides/client-reset.txt [hunk 8]
```diff
@@ -14,32 +14,81 @@ Client Resets - .NET SDK
...
--------
-When a reset is required, {+realm+} sends a  
-:dotnet-sdk:`ClientResetException <reference/Realms.Sync.Exceptions.ClientResetException.html>`. 
-You can add a listener on the 
-:dotnet-sdk:`Session.Error <reference/Realms.Sync.Session.html#Realms_Sync_Session_Error>` 
-event and handle the client reset by calling the 
+.. versionadded:: 10.11.1
+
+Starting in version 10.11.1, the .NET SDK provides the option to specify a client 
+reset strategy in your :dotnet-sdk:`FlexibleSyncConfiguration
```
Comment (nirinchev) [inline source/sdk/dotnet/advanced-guides/client-reset.txt:37]: Flexible sync only supports `ManualRecoveryHandler`. This is true for all SDKs, not just for .NET.

File: source/sdk/dotnet/advanced-guides/client-reset.txt [hunk 9]
```diff
@@ -14,32 +14,81 @@ Client Resets - .NET SDK
...
+----------------------------------
+When you specify a ``DiscardLocalResetHandler``, Realm 
+automatically performs a client reset with minimal code and minimal disruption 
+to your application workflow. This client reset mode restores your local 
+realm file to a syncable state without closing the realm and while maintaining  
+notifications.
+
+Be aware that this strategy *permanently deletes* all local unsynced changes 
+made since the last successful sync. To ensure local changes are incorporated,
+specify the :dotnet-sdk:`OnBeforeReset<TBD>` and
```
Comment (nirinchev) [inline source/sdk/dotnet/advanced-guides/client-reset.txt:54]: I don't think we should give vague advice like this in the docs. There's no trivial way for developers to do the merge and we're not explaining how they can do it, so we might as well not tell them to do it. A more actionable advice would be to use the `OnBeforeReset` callback to notify the user of the app that a reset is going to happen and `OnAfterReset` to notify them that the reset has completed.

File: source/examples/generated/dotnet/ClientResetExamples.codeblock.ManualClientReset.cs [hunk 10]
```diff
@@ -0,0 +1,39 @@
+private void SetupRealm()
+{
+    var fsConfig = new FlexibleSyncConfiguration(user);
+    fsConfig.OnSessionError = HandleSessionError;
```
Comment (nirinchev) [inline source/examples/generated/dotnet/ClientResetExamples.codeblock.ManualClientReset.cs:4]: OnSessionError is deprecated. Instead, users should use the `ClientResetHandler` property on the config and set it to `ManualRecoveryHandler`. Some details about it can be found in the changelog update here: https://github.com/realm/realm-dotnet/pull/2777/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4ed

File: source/sdk/dotnet/advanced-guides/client-reset.txt (hunk 8 above)
Comment (MongoCaleb) [reply source/sdk/dotnet/advanced-guides/client-reset.txt:37]: @nirinchev since I can set a FlexibleSyncConfiguration.ClientResetHandler to DiscardLocalResetHandler, should the SDK be updated with a build time error?

File: source/examples/generated/dotnet/ClientResetExamples.codeblock.ManualClientReset.cs (hunk 10 above)
Comment (MongoCaleb) [reply source/examples/generated/dotnet/ClientResetExamples.codeblock.ManualClientReset.cs:4]: @nirinchev can you help me parse this warning when I try to use:

`fsConfig.ClientResetHandler = new ManualRecoveryHandler()`

> ManualRecoveryHandler.ManualRecoveryHandler()' is obsolete: 'Use SyncConfigurationBase.OnSessionError in conjunction with SyncConfigurationBase.ClientResetHandler instead.'

File: source/sdk/dotnet/advanced-guides/client-reset.txt (hunk 8 above)
Comment (MongoCaleb) [reply source/sdk/dotnet/advanced-guides/client-reset.txt:37]: @nirinchev the SDK "lets" me use `DiscardLocalResetHandler` on `FlexibleSyncConfiguration`. So if that isn't supported, should the SDK be updated with a compile time error?

File: source/sdk/dotnet/advanced-guides/client-reset.txt (hunk 2 above)
Comment (MongoCaleb) [reply source/sdk/dotnet/advanced-guides/client-reset.txt:24]: I try to avoid the passive voice (or, rather, it is avoided by me), but in this case I feel like it's OK: the app is restarted...which could be an intentional restart by the user, or the system kills to app to free up memory, or the phone is rebooted for a software update, or...

Comment (MongoCaleb) [reply source/sdk/dotnet/advanced-guides/client-reset.txt:24]: Same with "Sync is terminated on the server" -- because there are many(?) reasons why this might happen, I think keeping it int he passive voice is OK. I'm 100% open to considering another angle on it!

File: source/sdk/dotnet/advanced-guides/client-reset.txt (hunk 3 above)
Comment (MongoCaleb) [reply source/sdk/dotnet/advanced-guides/client-reset.txt:25]: Is it uncommon?

File: source/sdk/dotnet/advanced-guides/client-reset.txt (hunk 8 above)
Comment (nirinchev) [reply source/sdk/dotnet/advanced-guides/client-reset.txt:37]: This is a runtime error currently - the plan is to eventually support it.

File: source/examples/generated/dotnet/ClientResetExamples.codeblock.ManualClientReset.cs (hunk 10 above)
Comment (nirinchev) [reply source/examples/generated/dotnet/ClientResetExamples.codeblock.ManualClientReset.cs:4]: You should use the constructor that takes an argument instead. The parameterless constructor is intended as a temporary measure to allow people to migrate without many code changes.

File: examples/dotnet/Examples/ClientResetExamples.cs [hunk 11]
```diff
@@ -7,68 +7,210 @@
...
-            config = new PartitionSyncConfiguration("myPart", user);
             //:hide-start:
             config.Schema = new[] { typeof(User) };
             //:hide-end:
             var realm = await Realm.GetInstanceAsync(config);
+        }
 
-            // :code-block-start: handle
-            Session.Error += (sender, err) =>
+        private void HandleBeforeResetCallbak(Realm beforeFrozen)
```
Comment (nirinchev) [inline examples/dotnet/Examples/ClientResetExamples.cs:108]: Typo: callbak -> callback

File: source/sdk/dotnet/advanced-guides/client-reset.txt [hunk 12]
```diff
@@ -14,32 +14,82 @@ Client Resets - .NET SDK
...
+To handle manual client resets, set the :dotnet-sdk:`OnSessionError <TBD>` callback
+of the ``FlexibleSyncConfiguration`` object. The callback takes two parameters:
+a :dotnet-sdk:`Session <reference/Realms.Sync.Session>` object and a 
+:dotnet-sdk:`SessionException <reference/Realms.Sync.Exceptions.SessionException.html>`
+object. 
+
+Within the handler, cast the SessionException to a 
+:dotnet-sdk:`ClientResetException <reference/Realms.Sync.Exceptions.ClientResetException.html>`,
+dispose the existing realm, and call the 
 :dotnet-sdk:`InitiateClientReset() <reference/Realms.Sync.Exceptions.ClientResetException.html#Realms_Sync_Exceptions_ClientResetException_InitiateClientReset>`
```
Comment (nirinchev) [inline source/sdk/dotnet/advanced-guides/client-reset.txt:87]: The wording in this section needs to be updated to match the code below

File: examples/dotnet/Examples/ClientResetExamples.cs (hunk 11 above)
Comment (MongoCaleb) [reply examples/dotnet/Examples/ClientResetExamples.cs:108]: That's the Danish way of spelling "callback". ;-D
